File: Parallel-Csharp/src/Parallel/Containers/QueueParallel.cs
namespace Parallel_Csharp.Parallel.Containers;

public class QueueParallel {
    // T1
    public Queue<char> GetQueueChar(){
        Queue<char> queue = new Queue<char>();
        for(int i=0; i<10; ++i){
            queue.Enqueue('c');
        }
        return queue;
    }
    
    // Convert Char to String
    public Queue<string> GetQueueString(){
        Queue<string> queue = new Queue<string>();
        for(int i=0; i<10; ++i){
            queue.Enqueue("aa");
        }
        return queue;
    } 
    
    // Convert Char to Int
    public Queue<int> GetQueueInt(){
        Queue<int> queue = new Queue<int>();
        for(int i=0; i<10; ++i){
            queue.Enqueue('c');
        }
        return queue;
    }
    
    // Convert Char to Double
    public Queue<double> GetQueueDouble(){
        Queue<double> queue = new Queue<double>();
        for(int i=0; i<10; ++i){
            queue.Enqueue('c');
        }
        return queue;
    }
    
    // T2
    public Queue<char> CopyAddChar(Queue<char> queue, char c){
        Queue<char> q = new Queue<char>(queue);
        q.Enqueue(c);
        return q;
    }
    
    // Convert Char to String
    public Queue<string> CopyAddString(Queue<string> queue, string c){
        Queue<string> q = new Queue<string>(queue);
        q.Enqueue(c);
        return q;
    }
    
    // Convert Char to Int
    public Queue<int> CopyAddInt(Queue<int> queue, int c){
        Queue<int> q = new Queue<int>(queue);
        q.Enqueue(c);
        return q;
    }
    
    // todo Convert Char to Double
    public Queue<double> CopyAddDouble(Queue<double> queue, double c){
        Queue<double> q = new Queue<double>(queue);
        q.Enqueue(c);
        return q;
    }
    
    // T3
    public Queue<char> ToLengthChar(Queue<char> queue, int length){
        int size = queue.Count;
        while(size < length){
            size++;
            queue.Enqueue('a');
        }
        return queue;
    }
    
    // Convert Char to String
    public Queue<string> ToLengthString(Queue<string> queue, int length){
        int size = queue.Count;
        while(size < length){
            size++;
            queue.Enqueue("aa");
        }
        return queue;
    }
    
    // Convert Char to Int
    public Queue<int> ToLengthInt(Queue<int> queue, int length){
        int size = queue.Count;
        while(size < length){
            size++;
            queue.Enqueue('a');
        }
        return queue;
    }
    
    // Convert Char to Double
    public Queue<double> ToLengthDouble(Queue<double> queue, int length){
        int size = queue.Count;
        while(size < length){
            size++;
            queue.Enqueue('a');
        }
        return queue;
    }
    
    // T4
    public string AddTwoChar(Queue<char> queue){
        char a = queue.Dequeue();
        char b = queue.First();
        queue.Enqueue(a);
        string str = "";
        str += a;
        str += b;
        return str;
    }
    
    // Convert Char to String
    public string AddTwoString(Queue<string> queue){
        string a = queue.Dequeue();
        string b = queue.First();
        queue.Enqueue(a);
        string str = "";
        str += a;
        str += b;
        return str;
    }
    
    // todo Convert Char to Int
    public int AddTwoInt(Queue<int> queue){
        int a = queue.Dequeue();
        int b = queue.First();
        queue.Enqueue(a);
        int str = 0;
        str += a;
        str += b;
        return str;
    }
    
    // todo Convert Char to Double
    public double AddTwoDouble(Queue<double> queue){
        double a = queue.Dequeue();
        double b = queue.First();
        queue.Enqueue(a);
        double str = 0.0;
        str += a;
        str += b;
        return str;
    }
    
    // T5
    public string AddAllChar(Queue<char> queue){
        string str = "";
        while(queue.Count != 0){
            str += queue.Dequeue();
        }
        return str;
    }
    
    // Convert Char to String
    public string AddAllString(Queue<string> queue){
        string str = "";
        while(queue.Count != 0){
            str += queue.Dequeue();
        }
        return str;
    }
    
    // todo Convert Char to Int
    public int AddAllInt(Queue<int> queue){
        int str = 0;
        while(queue.Count != 0){
            str += queue.Dequeue();
        }
        return str;
    }
    
    // Convert Char to Double
    public double AddAllDouble(Queue<double> queue){
        double str = 0;
        while(queue.Count != 0){
            str += queue.Dequeue();
        }
        return str;
    }
    
    // T6
    public Queue<char> ConcatQueueChar(Queue<char> queue1, Queue<char> queue2){
        Queue<char> queue = new Queue<char>(queue1);
        Queue<char> tmp = new Queue<char>(queue2);
        while(tmp.Count != 0){
            char c = tmp.Dequeue();
            queue.Enqueue(c);
        }
        tmp.Clear();
        return queue;
    }
    
    // Convert Char to String
    public Queue<string> ConcatQueueString(Queue<string> queue1, Queue<string> queue2){
        Queue<string> queue = new Queue<string>(queue1);
        Queue<string> tmp = new Queue<string>(queue2);
        while(tmp.Count != 0){
            string c = tmp.Dequeue();
            queue.Enqueue(c);
        }
        tmp.Clear();
        return queue;
    }
    
    // todo Convert Char to Int
    public Queue<int> ConcatQueueInt(Queue<int> queue1, Queue<int> queue2){
        Queue<int> queue = new Queue<int>(queue1);
        Queue<int> tmp = new Queue<int>(queue2);
        while(tmp.Count != 0){
            int c = tmp.Dequeue();
            queue.Enqueue(c);
        }
        tmp.Clear();
        return queue;
    }
    
    // todo Convert Char to Double
    public Queue<double> ConcatQueueDouble(Queue<double> queue1, Queue<double> queue2){
        Queue<double> queue = new Queue<double>(queue1);
        Queue<double> tmp = new Queue<double>(queue2);
        while(tmp.Count != 0){
            double c = tmp.Dequeue();
            queue.Enqueue(c);
        }
        tmp.Clear();
        return queue;
    }
    
    // T7
    public void DismissQueueChar(Queue<char> queue1, Queue<char> queue2){
        while(queue1.Count != 0 && queue2.Count != 0){
            char c1 = queue1.First();
            char c2 = queue2.First();
            if(c1 != c2) return;
            else{
                queue1.Dequeue();
                queue2.Dequeue();
            }
        }
    }
    
    // Convert Char to String
    public void DismissQueueString(Queue<string> queue1, Queue<string> queue2){
        while(queue1.Count != 0 && queue2.Count != 0){
            string c1 = queue1.First();
            string c2 = queue2.First();
            if(c1.CompareTo(c2) != 0) return;
            else{
                queue1.Dequeue();
                queue2.Dequeue();
            }
        }
    }
    
    // Convert Char to Int
    public void DismissQueueInt(Queue<int> queue1, Queue<int> queue2){
        while(queue1.Count != 0 && queue2.Count != 0){
            int c1 = queue1.First();
            int c2 = queue2.First();
            if(c1 != c2) return;
            else{
                queue1.Dequeue();
                queue2.Dequeue();
            }
        }
    }
    
    // Convert Char to Double
    public void DismissQueueDouble(Queue<double> queue1, Queue<double> queue2){
        while(queue1.Count != 0 && queue2.Count != 0){
            double c1 = queue1.First();
            double c2 = queue2.First();
            if(c1 != c2) return;
            else{
                queue1.Dequeue();
                queue2.Dequeue();
            }
        }
    }
    
    // T8
    public bool HeadTailEqualChar(Queue<char> queue){
        char c1 = queue.First();
        char c2 = queue.Last();
        if(c1 == c2) return true;
        else return false;
    }
        
    // Convert Char to String
    public bool HeadTailEqualString(Queue<string> queue){
        string c1 = queue.First();
        string c2 = queue.Last();
        if(c1.CompareTo(c2) == 0) return true;
        else return false;
    }
    
    // Convert Char to Int
    public bool HeadTailEqualInt(Queue<int> queue){
        int c1 = queue.First();
        int c2 = queue.Last();
        if(c1 == c2) return true;
        else return false;
    }
    
    // todo Convert Char to Double
    public bool HeadTailEqualDouble(Queue<double> queue){
        double c1 = queue.First();
        double c2 = queue.Last();
        if(c1 == c2) return true;
        else return false;
    }
    
    // T9
    public bool EqualQueueChar(Queue<char> queue1, Queue<char> queue2){
        if(queue1.Count != queue2.Count) return false;
        else return queue1.SequenceEqual(queue2);
    }
    
    // Convert Char to String
    public bool EqualQueueString(Queue<string> queue1, Queue<string> queue2){
        if(queue1.Count != queue2.Count) return false;
        else return queue1.SequenceEqual(queue2);
    }
    
    // Convert Char to Int
    public bool EqualQueueInt(Queue<int> queue1, Queue<int> queue2){
        if(queue1.Count != queue2.Count) return false;
        else return queue1.SequenceEqual(queue2);
    }
    
    // Convert Char to Double
    public bool EqualQueueDouble(Queue<double> queue1, Queue<double> queue2){
        if(queue1.Count != queue2.Count) return false;
        else return queue1.SequenceEqual(queue2);
    }
}

File: Parallel-Java/src/parallel/containers/QueueParallel.java
package parallel.containers;

import java.util.ArrayDeque;
import java.util.Arrays;

public class QueueParallel {
    // T1
    public ArrayDeque<Character> getQueueChar(){
        ArrayDeque<Character> queue = new ArrayDeque<>();
        for(int i=0; i<10; ++i){
            queue.add('c');
        }
        return queue;
    }

    // Convert Char to String
    public ArrayDeque<String> getQueueString(){
        ArrayDeque<String> queue = new ArrayDeque<>();
        for(int i=0; i<10; ++i){
            queue.add("aaa");
        }
        return queue;
    }
    
    // Convert Char to Int
    public ArrayDeque<Integer> getQueueInt(){
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        for(int i=0; i<10; ++i){
            queue.add(0);
        }
        return queue;
    }
    
    // Convert Char to Double
    public ArrayDeque<Double> getQueueDouble(){
        ArrayDeque<Double> queue = new ArrayDeque<>();
        for(int i=0; i<10; ++i){
            queue.add(0.0);
        }
        return queue;
    }

    // T2
    public ArrayDeque<Character> copyAddChar(ArrayDeque<Character> queue, Character c){
        ArrayDeque<Character> q = new ArrayDeque<>(queue);
        q.add(c);
        return q;
    }

    // Convert Char to String
    public ArrayDeque<String> copyAddString(ArrayDeque<String> queue, String c){
        ArrayDeque<String> q = new ArrayDeque<>(queue);
        q.add(c);
        return q;
    }
    
    // Convert Char to Int
    public ArrayDeque<Integer> copyAddInt(ArrayDeque<Integer> queue, Integer c){
        ArrayDeque<Integer> q = new ArrayDeque<>(queue);
        q.add(c);
        return q;
    }
    
    // Convert Char to Double
    public ArrayDeque<Double> copyAddDouble(ArrayDeque<Double> queue, Double c){
        ArrayDeque<Double> q = new ArrayDeque<>(queue);
        q.add(c);
        return q;
    }
    
    // T3
    public ArrayDeque<Character> toLengthChar(ArrayDeque<Character> queue, int length){
        int size = queue.size();
        while(size < length){
            size++;
            queue.add('a');
        }
        return queue;
    }

    // Convert Char to String
    public ArrayDeque<String> toLengthString(ArrayDeque<String> queue, int length){
        int size = queue.size();
        while(size < length){
            size++;
            queue.add("aa");
        }
        return queue;
    }

    // Convert Char to Int
    public ArrayDeque<Integer> toLengthInt(ArrayDeque<Integer> queue, int length){
        int size = queue.size();
        while(size < length){
            size++;
            queue.add(0);
        }
        return queue;
    }
    
    // Convert Char to Double
    public ArrayDeque<Double> toLengthDouble(ArrayDeque<Double> queue, int length){
        int size = queue.size();
        while(size < length){
            size++;
            queue.add(0.0);
        }
        return queue;
    }
    
    // T4
    public String addTwoChar(ArrayDeque<Character> queue){
        Character a = queue.remove();
        Character b = queue.peekFirst();
        queue.add(a);
        String str = "";
        str += a;
        str += b;
        return str;
    }

    // Convert Char to String
    public String addTwoString(ArrayDeque<String> queue){
        String a = queue.remove();
        String b = queue.peekFirst();
        queue.add(a);
        String str = "";
        str += a;
        str += b;
        return str;
    }
    
    // Convert Char to Int
    public Integer addTwoInt(ArrayDeque<Integer> queue){
        Integer a = queue.remove();
        Integer b = queue.peekFirst();
        queue.add(a);
        Integer str = 0;
        str += a;
        str += b;
        return str;
    }
    
    // Convert Char to Double
    public Double addTwoDouble(ArrayDeque<Double> queue){
        Double a = queue.remove();
        Double b = queue.peekFirst();
        queue.add(a);
        Double str = 0.0;
        str += a;
        str += b;
        return str;
    }
    
    // T5
    public String addAllChar(ArrayDeque<Character> queue){
        String str = "";
        while(!queue.isEmpty()){
            str += queue.remove();
        }
        return str;
    }
    
    // Convert Char to String
    public String addAllString(ArrayDeque<String> queue){
        String str = "";
        while(!queue.isEmpty()){
            str += queue.remove();
        }
        return str;
    }
    
    // Convert Char to Int
    public Integer addAllInt(ArrayDeque<Integer> queue){
        Integer str = 0;
        while(!queue.isEmpty()){
            str += queue.remove();
        }
        return str;
    }
    
    // Convert Char to Double
    public Double addAllDouble(ArrayDeque<Double> queue){
        Double str = 0.0;
        while(!queue.isEmpty()){
            str += queue.remove();
        }
        return str;
    }
    
    // T6
    public ArrayDeque<Character> concatQueueChar(ArrayDeque<Character> queue1, ArrayDeque<Character> queue2){
        ArrayDeque<Character> queue = new ArrayDeque<>(queue1);
        ArrayDeque<Character> tmp = new ArrayDeque<>(queue2);
        while(!tmp.isEmpty()){
            Character c = tmp.remove();
            queue.add(c);
        }
        tmp.clear();
        return queue;
    }
    
    // Convert Char to String
    public ArrayDeque<String> concatQueueString(ArrayDeque<String> queue1, ArrayDeque<String> queue2){
        ArrayDeque<String> queue = new ArrayDeque<>(queue1);
        ArrayDeque<String> tmp = new ArrayDeque<>(queue2);
        while(!tmp.isEmpty()){
            String c = tmp.remove();
            queue.add(c);
        }
        tmp.clear();
        return queue;
    }

    // Convert Char to Int
    public ArrayDeque<Integer> concatQueueInt(ArrayDeque<Integer> queue1, ArrayDeque<Integer> queue2){
        ArrayDeque<Integer> queue = new ArrayDeque<>(queue1);
        ArrayDeque<Integer> tmp = new ArrayDeque<>(queue2);
        while(!tmp.isEmpty()){
            Integer c = tmp.remove();
            queue.add(c);
        }
        tmp.clear();
        return queue;
    }
    
    // Convert Char to Double
    public ArrayDeque<Double> concatQueueDouble(ArrayDeque<Double> queue1, ArrayDeque<Double> queue2){
        ArrayDeque<Double> queue = new ArrayDeque<>(queue1);
        ArrayDeque<Double> tmp = new ArrayDeque<>(queue2);
        while(!tmp.isEmpty()){
            Double c = tmp.remove();
            queue.add(c);
        }
        tmp.clear();
        return queue;
    }

    // T7
    public void dismissQueueChar(ArrayDeque<Character> queue1, ArrayDeque<Character> queue2){
        while(!queue1.isEmpty() && !queue2.isEmpty()){
            Character c1 = queue1.peekFirst();
            Character c2 = queue2.peekFirst();
            if(c1 != c2) return;
            else{
                queue1.remove();
                queue2.remove();
            }
        }
    }
    
    // Convert Char to String
    public void dismissQueueString(ArrayDeque<String> queue1, ArrayDeque<String> queue2){
        while(!queue1.isEmpty() && !queue2.isEmpty()){
            String c1 = queue1.peekFirst();
            String c2 = queue2.peekFirst();
            if(c1.compareTo(c2) != 0) return;
            else{
                queue1.remove();
                queue2.remove();
            }
        }
    }
    
    // Convert Char to Int
    public void dismissQueueInt(ArrayDeque<Integer> queue1, ArrayDeque<Integer> queue2){
        while(!queue1.isEmpty() && !queue2.isEmpty()){
            Integer c1 = queue1.peekFirst();
            Integer c2 = queue2.peekFirst();
            if(c1 != c2) return;
            else{
                queue1.remove();
                queue2.remove();
            }
        }
    }
    
    // Convert Char to Double
    public void dismissQueueDouble(ArrayDeque<Double> queue1, ArrayDeque<Double> queue2){
        while(!queue1.isEmpty() && !queue2.isEmpty()){
            Double c1 = queue1.peekFirst();
            Double c2 = queue2.peekFirst();
            if(c1 != c2) return;
            else{
                queue1.remove();
                queue2.remove();
            }
        }
    }
    
    // T8
    public boolean headTailEqualChar(ArrayDeque<Character> queue){
        Character c1 = queue.peekFirst();
        Character c2 = queue.peekLast();
        if(c1 == c2) return true;
        else return false;
    }
    
    // Convert Char to String
    public boolean headTailEqualString(ArrayDeque<String> queue){
        String c1 = queue.peekFirst();
        String c2 = queue.peekLast();
        if(c1.compareTo(c2) == 0) return true;
        else return false;
    }
    
    // Convert Char to Int
    public boolean headTailEqualInt(ArrayDeque<Integer> queue){
        Integer c1 = queue.peekFirst();
        Integer c2 = queue.peekLast();
        if(c1 == c2) return true;
        else return false;
    }
    
    // Convert Char to Double
    public boolean headTailEqualDouble(ArrayDeque<Double> queue){
        Double c1 = queue.peekFirst();
        Double c2 = queue.peekLast();
        if(c1 == c2) return true;
        else return false;
    }
    
    // T9
    public boolean equalQueueChar(ArrayDeque<Character> queue1, ArrayDeque<Character> queue2){
        if(queue1.size() != queue2.size()) return false;
        else return Arrays.equals(queue1.toArray(), queue2.toArray());
    }

    // Convert Char to String
    public boolean equalQueueString(ArrayDeque<String> queue1, ArrayDeque<String> queue2){
        if(queue1.size() != queue2.size()) return false;
        else return Arrays.equals(queue1.toArray(), queue2.toArray());
    }
    
    // Convert Char to Int
    public boolean equalQueueInt(ArrayDeque<Integer> queue1, ArrayDeque<Integer> queue2){
        if(queue1.size() != queue2.size()) return false;
        else return Arrays.equals(queue1.toArray(), queue2.toArray());
    }
    
    // Convert Char to Double
    public boolean equalQueueDouble(ArrayDeque<Double> queue1, ArrayDeque<Double> queue2){
        if(queue1.size() != queue2.size()) return false;
        else return Arrays.equals(queue1.toArray(), queue2.toArray());
    }

    // NT1
    public void removeToEndChar(ArrayDeque<Character> queue1, Character ele){
        int size = queue1.size();
        for(int i = 0; i< size; ++i){
            if(ele != queue1.peekFirst()){
                Character t = queue1.remove();
                queue1.add(t);
            }
        }
    }

    public void removeUntilChar(ArrayDeque<Character> queue1, Character ele){
        while(!queue1.isEmpty() && queue1.peekFirst() != ele){
            queue1.remove();
        }
    }

    public void removeUntilLengthChar(ArrayDeque<Character> queue1, int length){
        if(queue1.size() <= length) return;
        while(queue1.size() > length){
            queue1.remove();
        }
    }
    
    // to String
    public void removeToEndString(ArrayDeque<String> queue1, String ele){
        int size = queue1.size();
        for(int i = 0; i< size; ++i){
            if(ele != queue1.peekFirst()){
                String t = queue1.remove();
                queue1.add(t);
            }
        }
    }

    public void removeUntilString(ArrayDeque<String> queue1, String ele){
        while(!queue1.isEmpty() && queue1.peekFirst() != ele){
            queue1.remove();
        }
    }

    public void removeUntilLengthString(ArrayDeque<String> queue1, int length){
        if(queue1.size() <= length) return;
        while(queue1.size() > length){
            queue1.remove();
        }
    }
    
    // to Int
    public void removeToEndInt(ArrayDeque<Integer> queue1, Integer ele){
        int size = queue1.size();
        for(int i = 0; i< size; ++i){
            if(ele != queue1.peekFirst()){
                Integer t = queue1.remove();
                queue1.add(t);
            }
        }
    }

    public void removeUntilInt(ArrayDeque<Integer> queue1, Integer ele){
        while(!queue1.isEmpty() && queue1.peekFirst() != ele){
            queue1.remove();
        }
    }

    public void removeUntilLengthInt(ArrayDeque<Integer> queue1, int length){
        if(queue1.size() <= length) return;
        while(queue1.size() > length){
            queue1.remove();
        }
    }
    
    // to Double
    public void removeToEndDouble(ArrayDeque<Double> queue1, Double ele){
        int size = queue1.size();
        for(int i = 0; i< size; ++i){
            if(ele != queue1.peekFirst()){
                Double t = queue1.remove();
                queue1.add(t);
            }
        }
    }

    public void removeUntilDouble(ArrayDeque<Double> queue1, Double ele){
        while(!queue1.isEmpty() && queue1.peekFirst() != ele){
            queue1.remove();
        }
    }

    public void removeUntilLengthDouble(ArrayDeque<Double> queue1, int length){
        if(queue1.size() <= length) return;
        while(queue1.size() > length){
            queue1.remove();
        }
    }

}
